File: src/HARD/45. Jump Game II.ts
// 09/08/2020 HARD

// https://leetcode.com/problems/jump-game-ii/


// function jump(nums: number[]): number {
//   const len = nums.length;
//   if (!len) return 0;
//   const dp: number[] = [];
//   dp[0] = 0;
//   for (let i = 1; i < len; i++) {
//     let minStep = Number.MAX_SAFE_INTEGER;
//     for (let j = 0; j < i; j++) {
//       if (nums[j] >= i - j) minStep = Math.min(minStep, dp[j]);
//     }
//     dp[i] = minStep + 1;
//   }
//   return dp[len-1];
// };


// Solution: Greedy
function jump(nums: number[]): number {
  const len = nums.length;
  if (!len) return 0;
  let cur = 0, last = 0, count = 0;
  for (let i = 0; i < len - 1; i++) {
    cur = Math.max(cur, i + nums[i]);
    if (last === i) {
      last = cur;
      count++;
    }
  }
  return count;
};

File: src/MEDIUM/45. Jump Game II.cpp
// 05/14/2021 MEDIUM

// https://leetcode-cn.com/problems/jump-game-ii/

/*
Given an array of non-negative integers nums, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Your goal is to reach the last index in the minimum number of jumps.

You can assume that you can always reach the last index.

 

Example 1:

Input: nums = [2,3,1,1,4]
Output: 2
Explanation: The minimum number of jumps to reach the last index is 2. 
Jump 1 step from index 0 to 1, then 3 steps to the last index.
Example 2:

Input: nums = [2,3,0,1,4]
Output: 2
 

Constraints:

1 <= nums.length <= 1000
0 <= nums[i] <= 105

*/
#include <vector>
#include <algorithm>
using namespace std;
const int INF = 1e3 + 1;
class Solution {
  public:
    int jump(vector<int>& nums) {
      int n = nums.size();
      vector<int> memo(n, INF);
      memo[n - 1] = 0;
      return dfs(0, memo, nums);
    }
    int dfs(int index, vector<int>& memo, vector<int>& nums) {
      if (index >= nums.size()) return INF;
      if (memo[index] != INF) return memo[index];
      for (int i = 1; i <= nums[index]; ++i) {
        memo[index] = min(memo[index], dfs(index + i, memo, nums) + 1);
      }
      return memo[index];
    }
};